iis-wf/iis-wf-import/src/main/java/eu/dnetlib/iis/wf/importer/infospace/ImportInformationSpaceJobUtils.java

```diff
@@ -0,0 +1,80 @@
+     */
+    public static JavaRDD<IdentifierMapping> produceObjectStoreId(JavaRDD<eu.dnetlib.dhp.schema.oaf.Dataset> sourceDataset,
+                                                                  JavaRDD<eu.dnetlib.dhp.schema.oaf.OtherResearchProduct> sourceOtherResearchProduct,
+                                                                  JavaRDD<eu.dnetlib.dhp.schema.oaf.Publication> sourcePublication,
+                                                                  JavaRDD<eu.dnetlib.dhp.schema.oaf.Software> sourceSoftware,
+                                                                  JavaRDD<IdentifierMapping> dedupRelation,
+                                                                  SparkSession spark) {
+        Dataset<eu.dnetlib.dhp.schema.oaf.Dataset> sourceDatasetDS = spark.createDataset(
+                sourceDataset.rdd(), Encoders.bean(eu.dnetlib.dhp.schema.oaf.Dataset.class));
+        Dataset<eu.dnetlib.dhp.schema.oaf.OtherResearchProduct> sourceOtherResearchProductDS = spark.createDataset(
+                sourceOtherResearchProduct.rdd(), Encoders.bean(eu.dnetlib.dhp.schema.oaf.OtherResearchProduct.class));
+        Dataset<eu.dnetlib.dhp.schema.oaf.Publication> sourcePublicationDS = spark.createDataset(
+                sourcePublication.rdd(), Encoders.bean(eu.dnetlib.dhp.schema.oaf.Publication.class));
+        Dataset<eu.dnetlib.dhp.schema.oaf.Software> sourceSoftwareDS = spark.createDataset(
+                sourceSoftware.rdd(), Encoders.bean(eu.dnetlib.dhp.schema.oaf.Software.class));
+
+        Column idIsNotDedup = not(col("id").like("%dedup%"));
+        Column oidIsResultType = col("oid").like(InfoSpaceConstants.ROW_PREFIX_RESULT + "%");
+
+        Dataset<Row> resultIdMapDF = sourceDatasetDS.select(col("id"), explode(col("originalId")).as("oid"))
+                .union(sourceOtherResearchProductDS.select(col("id"), explode(col("originalId")).as("oid")))
+                .union(sourcePublicationDS.select(col("id"), explode(col("originalId")).as("oid")))
+                .union(sourceSoftwareDS.select(col("id"), explode(col("originalId")).as("oid")))
+                .where(idIsNotDedup)
+                .where(oidIsResultType)
+                .distinct();
+
+        Dataset<Row> dedupRelationDF = new AvroDatasetSupport(spark).toDF(
+                spark.createDataset(dedupRelation.rdd(), Encoders.kryo(IdentifierMapping.class)), IdentifierMapping.SCHEMA$);
+
+        Column dedupMappingNotMatched = col("newId").isNull().and(col("id").isNotNull());
+
+        Dataset<Row> identifierMappingDF = resultIdMapDF
+                .join(dedupRelationDF, col("id").equalTo(col("originalId")), "left_outer")
+                .select(
+                        when(dedupMappingNotMatched, col("id"))
+                                .otherwise(col("newId")).as("newId"),
+                        col("oid").as("originalId")
+                );
```

Comment:
It is quite difficult to verify the (rather simple) logic behind this piece of code, possibly because of the brackets I need to count in order to know on which object each method is called. Maybe new lines or using temp variables could help a little bit to increase the readability?

One another thing which could improve the readability: why using condition with inclusive negation here? Usually it is much easier to rely on `dedupMappingMatched` condition instead of `dedupMappingNotMatched` especially when you have `otherwise` as a next operator in a sequence.

Reply:
There are only two brackets grouped at max, so it's not that much ;-) `when`/`otherwise` is like simple 'if' here so there's not much place for increasing readability or using temp variables. I've put the whole statement on one line, I think this helps a bit.